tools/rccl_replayer/rcclReplayer.cpp

```diff
@@ -0,0 +1,379 @@
+        recvNumBytes = numGlobalRanks * sendNumBytes;
+    } else if (taskInfo.funcType == ncclCollReduceScatter || taskInfo.funcType == ncclCollScatter) {
+        recvNumBytes = taskInfo.count * DataTypeToBytes(taskInfo.datatype); 
+        sendNumBytes = numGlobalRanks * recvNumBytes; 
+    } else if (taskInfo.funcType == ncclCollAllToAll) {
+        sendNumBytes = numGlobalRanks * taskInfo.count * DataTypeToBytes(taskInfo.datatype);
+        recvNumBytes = sendNumBytes;
+    } else {
+        sendNumBytes = taskInfo.count * DataTypeToBytes(taskInfo.datatype);
+        recvNumBytes = sendNumBytes;
+    }
+    return std::make_pair(sendNumBytes, recvNumBytes);
+}
+
+void ExecuteCollective(TaskInfo task, ncclComm_t comm, hipStream_t stream, const void *sendbuff, void *recvbuff) {
+
+    int funcTypeValue = (int)task.funcType;
+
+    switch (funcTypeValue) {
+        case ncclCollAllGather:
+            NCCLCHECK(ncclAllGather(sendbuff, recvbuff, task.count, task.datatype, comm, stream));
+            break;
+        case ncclCollAllReduce:
+            NCCLCHECK(ncclAllReduce(sendbuff, recvbuff, task.count, task.datatype, task.op, comm, stream));
+            break;
+        case ncclCollBroadcast:
+            NCCLCHECK(ncclBroadcast(sendbuff, recvbuff, task.count, task.datatype, task.root, comm, stream));
+            break;
+        case ncclCollReduce:
+            NCCLCHECK(ncclReduce(sendbuff, recvbuff, task.count, task.datatype, task.op, task.root, comm, stream));
+            break;
+        case ncclCollReduceScatter:
+            NCCLCHECK(ncclReduceScatter(sendbuff, recvbuff, task.count, task.datatype, task.op, comm, stream));
+            break;
+        case ncclCollGather:
+            NCCLCHECK(ncclGather(sendbuff, recvbuff, task.count, task.datatype, task.root, comm, stream));
+            break;
+        case ncclCollScatter:
+            NCCLCHECK(ncclScatter(sendbuff, recvbuff, task.count, task.datatype, task.root, comm, stream));
+            break;
+        case ncclCollAllToAll:
+            NCCLCHECK(ncclAllToAll(sendbuff, recvbuff, task.count, task.datatype, comm, stream));
+            break;
+        case ncclCollSend:
+            NCCLCHECK(ncclSend(sendbuff, task.count, task.datatype, task.root, comm, stream));
+            break;
+        case ncclCollRecv:
+            NCCLCHECK(ncclRecv(recvbuff, task.count, task.datatype, task.root, comm, stream));
+            break;
+        default:
+            printf("Error: unsupported collective\n");
+            exit(1);
+    }
+}
+
+void ReplayRccl(GroupCall& groupCall, std::vector<ncclComm_t> comms, std::vector<hipStream_t> streams,
+                                                                     int const localGpuOffset, int const numGpusPerMpiRank, int const firstGlobalRank, int const numGlobalRanks) {
+    std::vector<TaskInfo>& tasks = groupCall.rankData[localGpuOffset].tasks;
+
+    void* sendbuff[numGpusPerMpiRank];
+    void* recvbuff[numGpusPerMpiRank];
```

Comment:
This looks wrong still.  Each task needs its own send/recv buffer for each GPU.
This should be a vector<vector<void*>> sendbuff(numGpusPerMpiRank), with the inner vector fluxuating with the number of tasks for that rank in the group call.